OpenSim/Actuators/Test/testMuscles.cpp

```diff
@@ -1108,10 +958,172 @@ void testDelp1990Muscle()
+    double cosphi = SimTK::NaN;
+    double flm = SimTK::NaN;
+    double a = SimTK::NaN;
+
+    double fvm = 1.5;
+
+    assert(!muscle.get_ignore_tendon_compliance());
+
+    SimTK::RowVector row(labels.size(), SimTK::NaN);
+    for (int i = 0; i < N; ++i) {
+        fiberLength += dl;
+        s.setTime(fiberLength);
+        muscle.setFiberLength(s, fiberLength);
+        muscle.getModel().realizeDynamics(s);
+
+        vmt = muscle.getSpeed(s);
+
+        tendonForce = muscle.getTendonForce(s);
 
+        a = muscle.getActivation(s);
+        
+        
+        flm = muscle.getActiveForceLengthMultiplier(s);
+        //fvm = muscle.getForceVelocityMultiplier(s);
+
+        cosphi = muscle.getCosPennationAngle(s);
+
+        //vm = (vmt-muscle.getPennationModel().calcTendonVelocity(s))*cosphi;
+
+
+        activeFiberForce = a*fiso*flm*fvm;
+            
+        passiveFiberForce = muscle.getPassiveFiberForce(s);
+
+        row[0] = fiberLength; // muscle.getFiberLength(s);
+        row[1] = muscle.getLength(s);
+        row[2] = tendonForce;
+        row[3] = (activeFiberForce + passiveFiberForce)*cosphi;
+        row[4] = muscle.getActivation(s);
+        row[5] = activeFiberForce;
+        row[6] = passiveFiberForce;
+        row[7] = row[3] - row[2];
+
+        forcesVsFiberLengthTable.appendRow(s.getTime(), row);
+    }
+
+    std::string fileName = "forcesVsFiberLength_"
+        + std::to_string(a) + ".sto";
+
+    STOFileAdapter::write(forcesVsFiberLengthTable, fileName);
+}
+
+void testMuscleEquilibriumSolve(const Model& model, const Storage& statesStore)
+{
+    // Get the muscle to test
+    const Muscle& muscle = model.getMuscles()[0];
+
+    if (dynamic_cast<const Millard2012AccelerationMuscle*>(&muscle)) {
+        // Millard2012AccelerationMuscle is not an Equilibrium-based Muscle
+        return;
+    }
+
+    // Load input data as StatesTrajectory used to perform the Analysis
+    auto statesTraj = StatesTrajectory::createFromStatesStorage(
+        model, statesStore, true, false);
+    size_t nstates = statesTraj.getSize();
+
+    // muscle active, passive, total muscle and tendon force quantities
+    double af, pf, mf, tf, nfl, fv;
+    af = pf = mf = tf = nfl = fv = SimTK::NaN;
+
+    // Tolerance for muscle equilibrium solution 
+    const double equilTol = muscle.getMaxIsometricForce()*SimTK::SqrtEps;
+
+    // The maximum acceptable change in force between two contiguous states
+    const double maxDelta = muscle.getMaxIsometricForce() / 2;
+
+    const double dAct = 0.05;
+    double activation = 0.01;
+
+    SimTK::State s = model.getWorkingState();
+    // Independently compute the active fiber force at every state
+    for (int i = 0; i < nstates; ++i) {
+        s = statesTraj[i];
+
+        activation = 0.01;
+        // test a full sweep of default activations at each state
+        for (int j = 0; j*dAct <= 1.0; ++j) {
+            activation = (!j)*activation + j*dAct;
```

Comment:
The logic here is unclear. Some muscles have a minimum allowable activation; presumably L1076 is meant to address that constraint. This code will break if `dAct < min_allowable_activation`, which is not obvious from the implementation. I suggest something more straightforward:
```cpp
const double minActivation = muscle.getMinimumActivation();
const int nSteps = 20;
const double dAct = (1.0 - minActivation) / nSteps;
for (int j = 0; j <= nSteps; ++j) {
    activation = std::min( minActivation + j*dAct, 1.0); //avoid exceeding 1 due to roundoff
    ...
}
```